src/Dictionaries/CacheDictionary.cpp

```diff
@@ -609,10 +609,17 @@ void CacheDictionary<dictionary_key_type>::update(CacheDictionaryUpdateUnitPtr<d
                 {
                     auto fetched_key_from_source = keys_extracted_from_block[i];
 
-                    not_found_keys.erase(fetched_key_from_source);
-                    update_unit_ptr->requested_keys_to_fetched_columns_during_update_index[fetched_key_from_source] = found_keys_size;
-                    found_keys_in_source.emplace_back(fetched_key_from_source);
-                    ++found_keys_size;
+                    /// The underlying source can have duplicates,
+                    /// but if we will process same key twice it will incorrectly account found_keys_size,
+                    /// and also overflows DictCacheKeysRequestedMiss.
+                    ///
+                    /// So only the first row with the same key will be processed.
+                    if (not_found_keys.erase(fetched_key_from_source))
```

Comment:
It probably breaks invariant of insertColumnsForKeys method of ICacheDictionaryStorage, number of columns rows  should equal to number of found keys, and they should be matched.

Reply:
Maybe better to remove found_keys_size here, instead use found_keys_in_source.size(), but at the end of function add such lines
```
size_t found_keys_size = requested_keys_size - not_found_keys.size()
ProfileEvents::increment(ProfileEvents::DictCacheKeysRequestedMiss, not_found_keys);
ProfileEvents::increment(ProfileEvents::DictCacheKeysRequestedFound, found_keys_size);
ProfileEvents::increment(ProfileEvents::DictCacheRequests);
```

Reply:
Good idea, done